### ISSUE TITLE
Out-of-memory

### ISSUE THREAD

**Issue by thangld201** [sai-prasanna/bert-experiments#2]
Hi, I'm using a 40GB VRAM memory GPU but get out-of-memory error. The error seems to come from `src/model_bert.py`, Line 339 (alpha_f):

```
            _, attention_probs, value_layer = self_outputs
            output_head_weights = self.output.dense.weight.view(self.self.num_attention_heads, self.self.attention_head_size, -1)
            proj_up_value_layer = torch.einsum("ijkl,jld->ijkd", value_layer, output_head_weights)
            alpha_f =  torch.einsum("ijkl,ijld->ijkld", attention_probs, proj_up_value_layer) # <---- OUT_OF_MEMORY
            outputs = (attention_output, attention_probs, alpha_f)
```

These lines do not exist in the original Transformers lib implementation. Would you mind explaining what alpha_f means here ? @annargrs  @sai-prasanna @arumshisky

**Comment by sai-prasanna** [comment]
Hi Thang LD, it has been a long while since I worked on this project. But I looked it up, it is for computing weight normed attention. If I am not mistaken it's the alpha term highlighted in this paper: https://arxiv.org/abs/2004.10102.

For your memory error, take a look at `experiment_analysis/head_classifiication_data_gen.ipynb`. I use `torch.no_grad()` to avoid computing gradients for this computation and I also delete the variable to clear the GPU memory everytime.

If you have any further doubts let me know. I will try to recollect things and help you out!

**Comment by thangld201** [comment]
@sai-prasanna Thanks for your swift response!

At my best (with a 40GB GPU), I can only use `max_seqlen=128` (same as repo) and  `batch size=4` with `torch.no_grad()`. It was weird because default batch size for every script in the repo (and paper) was `64` , and this `alpha_f` was set to be computed in every  `bert.forward()` (though not always used).

If I use any `batch_size>4`, it would break in a single batch. I also double-checked the implementation from https://arxiv.org/abs/2004.10102 and saw that it was similar. I have no idea what is wrong here (literally just cloned the repo ..).

At least `alpha_f` was not used in most scripts, so I commented out the part about it for now. I will leave this issue here and see if we can find the cause later.
